runtime/include/chpl-comm.h

```diff
@@ -508,6 +512,53 @@ typedef struct _chpl_commDiagnostics {
   uint64_t execute_on_nb;
 } chpl_commDiagnostics;
 
+typedef struct _chpl_atomic_commDiagnostics {
```

Comment:
I definitely like having the internal counters implemented just once and in a high-performance way, but I'm not fond of that being done in `chpl-comm.h` which is where one would expect to find the interface to the comm layer rather than its exposed internals.  Would you object to putting this in, say, a `chpl-comm-internal.h` file or similar, and `#include`'ing that in the various comm layers, in addition to `chpl-comm.h`?

Reply:
Yeah, I didn't love it either, but the reason I put it here is to have the 2 structs and the copying/init code right next to each other. That way if we change the `chpl_commDiagnostics` "API" to the module code, we'll be more likely to remember to update the atomic API too.

Reply:
We could code a generator macro to declare the struct type and members, with the struct tag and member type being parameters.  Then just invoke this one way to build the public type in `chpl-comm.h` and another way to build the private one in `chpl-comm-internal.h`.  (I wouldn't mind the generator being in the public .h file.)

Oh, wait, that won't work.  Codegen via LLVM wouldn't be able to deal with it, for the usual reason.  Sigh.

All right.  Would you mind at least moving the internal struct decl down to the bottom of the file, say, with a header comment saying something like "stuff below here isn't part of the comm interface"?  (Maybe combine this with the generator macro idea if you don't like separating the two structs and their should-correspond-with-each-other member lists.)

Reply:
I don't think the generator is an issue for llvm (otherwise our atomics/network atomics wouldn't work) 

I _think_ the problem is for function like macros

Reply:
Hmm!  Now I'm confused.  If the LLVM issue was for function-like macros it seems like it would affect ugni's `CHPL_COMM_IMPL_REG_MEM_FREE()`.